### ISSUE TITLE
utils: All computeds are methods

### ISSUE THREAD

**Issue by buschtoens** [ember-decorators/ember-decorators#315]
https://github.com/ember-decorators/ember-decorators/blob/3d9327eda2c0908332486fed755ba2ac563fb1bc/packages/utils/addon/computed.js#L13-L26

> All computeds are methods

Are they? I am currently working on updating `ember-concurrency-decorators` to the stage 2 version of `@ember-decorators/utils`: https://github.com/machty/ember-concurrency-decorators/pull/28

I used to be able to create decorators with `computedDecoratorWithRequiredParams` that would create a `computed` from a _field_, like:

```js
class ObjectWithTaskDefaultValue extends EmberObject {
  @task
  *task() {
    return yield 'foo';
  };

  @lastValue('task')
  value = 'default value';
}
```

This now breaks with the following error:

```
TypeError: A method descriptor can't have a .initializer property.
    at _disallowProperty (http://localhost:4200/assets/tests.js:5398:92)
    at _toElementDescriptor (http://localhost:4200/assets/tests.js:5390:1006)
    at _toElementFinisherExtras (http://localhost:4200/assets/tests.js:5392:68)
    at _decorateElement (http://localhost:4200/assets/tests.js:5382:339)
    at http://localhost:4200/assets/tests.js:5378:366
    at Array.forEach (<anonymous>)
    at _decorateClass (http://localhost:4200/assets/tests.js:5378:244)
    at _decorate (http://localhost:4200/assets/tests.js:5360:182)
    at Object.<anonymous> (http://localhost:4200/assets/tests.js:5452:40)
    at runTest (http://localhost:4200/assets/test-support.js:5569:30)
```

Basically, I would expect the following code to work:

```js
export default computedDecoratorWithRequiredParams(function(
  { initializer },
  [task]
) {
  return computed(`${task}.lastSuccessful.value`, function() {
    const lastInstance = get(this, `${task}.lastSuccessful`);

    if (lastInstance) {
      return get(lastInstance, 'value');
    }

    if (initializer) {
      return initializer.call(this);
    }
  });
},
'lastValue');
```

**Comment by pzuraq** [comment]
Yeah, I think that's incorrect. The reason it works as is is because our finishers actually redefine the property, but if you tried to do this the way it _should_ be done (and the way it will be done in the future), the final descriptor that is returned from the decorator is exactly what will be applied to the class.

If you return a type `field` (or `property` or whatever it is), then you will get a field, not a getter.

In your case, you are transform a field definition _into_ a getter method. You are using the initializer of the field in the method itself here, and the initializer is _not_ being used as an actual initializer, so what you need to do is to delete the initializer before you pass the descriptor on:

```js
export default computedDecoratorWithRequiredParams(function(
  desc,
  [task]
) {
  let { initializer } = desc;
  delete desc.initializer; // or set it to null/undefined if that works

  return computed(`${task}.lastSuccessful.value`, function() {
    const lastInstance = get(this, `${task}.lastSuccessful`);

    if (lastInstance) {
      return get(lastInstance, 'value');
    }

    if (initializer) {
      return initializer.call(this);
    }
  });
},
'lastValue');
```

**Comment by buschtoens** [comment]
Make sense! But in that case the deletion of the `initializer` would have to happen somewhere in [`computedDecoratorInner`](https://github.com/ember-decorators/ember-decorators/blob/3d9327eda2c0908332486fed755ba2ac563fb1bc/packages/utils/addon/computed.js#L22-L67) already, since it would be to late when the `finisher` and thus the passed `fn` is executed.

Do you consider this out of scope of `computedDecoratorWithRequiredParams` and should I rather built a custom utility function? I could imagine that this sort of syntax might become popular though.

**Comment by pzuraq** [comment]
Yes, I consider this out of scope. Currently working on the decorators RFC for Ember and the utils will likely not be necessary any more once we get decorators into Ember proper I think we should minimize their usage as much as possible.

**Comment by pzuraq** [comment]
To be honest I think that'd be worse 😅 what I'm going to propose in RFC is that `computed` itself _directly_ returns a decorator. This will make the following Just Work as a decorator:

```js
function(desc) {
  let { initializer } = desc;
  delete desc.initializer; // or set it to null/undefined if that works

  return computed(`${task}.lastSuccessful.value`, function() {
    const lastInstance = get(this, `${task}.lastSuccessful`);

    if (lastInstance) {
      return get(lastInstance, 'value');
    }

    if (initializer) {
      return initializer.call(this);
    }
  });
}
```

**Comment by pzuraq** [comment]
Sorry, didn't mean to close! But yeah, basically if this RFC goes through, the next version of `ember-decorators` will likely solve this issue in a much cleaner way.

If you can think of some way to modify the existing utility functions in the meantime that gets you where you need to be, I'm open to merging it to unblock you and `ember-concurrency-decorators`, but like I said I generally want to discourage use of the `@ember-decorators/utils` package.

**Comment by buschtoens** [comment]
Got ya. What about compatibility with legacy apps then? One big appeal of using `utils` is that most inconsistencies are abstracted away from you.

**Comment by pzuraq** [comment]
I'm _pretty_ sure we will be able to make the `@ember-decorators` version of `computed` work the same way in v4, so you'll be able to define macros using the exact same syntax as you would with normal computed properties.

**Comment by buschtoens** [comment]
Sweet! I'll ponder on a solution in the meantime. Thanks!

**Comment by buschtoens** [comment]
That does the trick 😊 

```js
export default function lastValue(taskName) {
  assert(
    `ember-concurrency-decorators: @lastValue expects a task name as the first parameter.`,
    typeof taskName === 'string'
  );

  return decorator(desc => {
    const { initializer } = desc;
    delete desc.initializer;

    return computedDecorator(() =>
      computed(`${taskName}.lastSuccessful`, function() {
        const lastInstance = get(this, `${taskName}.lastSuccessful`);

        if (lastInstance) {
          return get(lastInstance, 'value');
        }

        if (initializer) {
          return initializer.call(this);
        }
      })
    )(desc);
  });
}
```
